src/Core/BinCodec.fs

```diff
@@ -0,0 +1,62 @@
+﻿namespace B2R2
+
+/// Little Endian Base 128 Codec
+module LEB128 =
+  [<Literal>]
+  let max32LEB128Length = 5
+  [<Literal>]
+  let max64LEB128Length = 10
+
+  let inline private decodeUnsignedInt (bytes: byte []) (castValue: byte -> 'T) maxLength =
+    let rec decodeLoop offset value currentByte length =
+      let newValue = value ||| (castValue (currentByte &&& 0x7fuy) <<< (offset * 7))
+      if currentByte &&& 0x80uy <> 0uy && offset = length - 1 then
+        Error "LEB128 Overflow"
+      else
+        if currentByte &&& 0x80uy = 0uy then
+          Ok (newValue, uint8 (offset + 1))
+        else
+          decodeLoop (offset + 1) newValue bytes.[offset + 1] length
+    if bytes.Length = 0 then
+      Error "Invalid buffer length"
+    else
+      let len = if bytes.Length > maxLength then maxLength else bytes.Length
+      decodeLoop 0 (castValue 0uy) bytes.[0] len
+
+  let inline private decodeSignedInt (bytes: byte []) (castValue: byte -> 'T) maxLength signExtension =
+    let rec decodeLoop offset value bits currentByte length =
+      let newValue = value ||| (castValue (currentByte &&& 0x7fuy) <<< bits)
+      if currentByte &&& 0x80uy <> 0uy && offset = length - 1 then
+        Error "LEB128 Overflow"
+      else
+        if currentByte &&& 0x80uy = 0uy then
+          let finalValue =
+            if currentByte &&& 0x40uy <> 0uy then
+              let shiftOffset = if offset < (maxLength - 1) then offset + 1 else offset
+              signExtension <<< (7 * (shiftOffset)) ||| newValue
+            else
+              newValue
+          Ok (finalValue, uint8 (offset + 1))
+        else
+          decodeLoop (offset + 1) newValue (bits + 7) bytes.[offset + 1] length
+    if bytes.Length = 0 then
+      Error "Invalid buffer length"
```

Comment:
Using the `Result` type is good for F#, but can be bad for interoperability.  I think you should simply return a value and raise an exception using a function like `invalidArg`: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/exception-handling/the-invalidarg-function.